File: frontend/src/types/mission.ts
// Legacy Mission Protocol status (for MissionProtocolDraft)
export type MissionProtocolStatus = "draft" | "in_progress" | "review" | "complete";

// New mission status aligned with backend API
export type MissionStatus = "draft" | "queued" | "in_progress" | "completed" | "blocked" | "cancelled" | "validation_failed";

export type QualityGateName =
  | "research_statement"
  | "evidence_links"
  | "contradictions_resolved"
  | "synthesis_quality"
  | "traceability";

export type QualityGateStatus = "pending" | "pass" | "fail";

export interface ResearchStatement {
  topic: string;
  objective: string;
  scope: string;
  audience?: string | null;
  methodology?: string | null;
  success_metrics: string[];
  risks: string[];
}

export interface KeyQuestion {
  question: string;
  status: "open" | "researching" | "answered";
  answer?: string | null;
  confidence?: number | null;
  owner?: string | null;
}

export interface Evidence {
  evidence_id: string;
  source: string;
  summary: string;
  chunk_id?: string | null;
  insight_id?: string | null;
  source_type?: string | null;
  relevance_score?: number | null;
  tags: string[];
}

export interface Synthesis {
  key_insights: string[];
  surprising_findings: string[];
  contradictory_information: string[];
  contradiction_resolutions: string[];
  recommendations: string[];
  next_steps: string[];
}

export interface QualityCheckpoint {
  gate: QualityGateName;
  status: QualityGateStatus;
  notes?: string | null;
  validated_by?: string | null;
  validated_at?: string | null;
}

export interface MissionProtocolDraft {
  mission_id: string;
  version: string;
  title?: string | null;
  summary?: string | null;
  project_id?: string | null;
  status: MissionProtocolStatus;
  owner?: string | null;
  research_statement?: ResearchStatement | null;
  key_questions: KeyQuestion[];
  synthesis?: Synthesis | null;
  evidence: Evidence[];
  quality_checkpoints: QualityCheckpoint[];
  tags: string[];
  created_at?: string | null;
  updated_at?: string | null;
}

export interface Mission {
  id: string;
  project_id: string | null;
  mission_data: MissionProtocolDraft;
  quality_gates?: Record<string, Record<string, unknown>> | null;
  status?: MissionProtocolStatus | null;
  completion_percentage?: number | null;
  created_at: string;
  updated_at: string;
}

export interface MissionCreatePayload {
  project_id: string;
  mission_data: MissionProtocolDraft;
  quality_gates?: Record<string, Record<string, unknown>>;
  status?: MissionProtocolStatus;
}

export type MissionUpdatePayload = Partial<MissionCreatePayload>;

export interface QualityGateReport {
  mission_id: string;
  protocol_mission_id: string;
  evaluated_at: string;
  all_passed: boolean;
  failing_gates: string[];
  gates: Record<
    QualityGateName,
    {
      gate: QualityGateName;
      status: QualityGateStatus;
      blocking: boolean;
      details?: string | null;
      evaluated_at: string;
      metadata?: Record<string, unknown> | null;
    }
  >;
}

export interface EvidenceLinkPayload {
  evidence_id: string;
  chunk_id: string;
  insight_id?: string;
}

// ============================================
// API Mission types (aligned with backend)
// ============================================

/**
 * Mission entity from the API (B16.1+ schema)
 */
export type ResearchDepth = "baseline" | "deep" | "alpha";

export interface MissionReference {
  title: string;
  [key: string]: unknown;
}

/** Mission-authoring fields consumed by DeepSearch's contract compiler (T40.1/T40.2). */
export interface MissionAuthoringFields {
  background?: string | null;
  focus?: string | null;
  references?: MissionReference[] | null;
  required_entities?: string[] | null;
  excluded_entities?: string[] | null;
  expected_output_schema?: Record<string, unknown> | null;
  coverage_thresholds?: Record<string, unknown> | null;
  validation_thresholds?: Record<string, unknown> | null;
  deliverable_format?: string | null;
  max_loops?: number | null;
  min_loops?: number | null;
  constraints?: string[] | null;
}

export interface ApiMission extends MissionAuthoringFields {
  id: string;
  project_id: string | null;
  project_name: string | null;
  mission_id: string;
  title: string;
  objective: string;
  success_criteria: string[];
  context: Record<string, unknown>;
  deliverables: string[];
  research_phases: Record<string, unknown>;
  tags: string[];
  metadata: Record<string, unknown>;
  research_depth: ResearchDepth;
  status: MissionStatus;
  queued_at: string | null;
  started_at: string | null;
  completed_at: string | null;
  deepsearch_job_id: string | null;
  execution_metadata: Record<string, unknown>;
  result_document_ids: string[];
  result_report_id: string | null;
  result_markdown: string | null;
  result_protocol: Record<string, unknown> | null;
  error_message: string | null;
  created_at: string;
  updated_at: string;
  created_by: string | null;
}

export interface ApiMissionCreate extends MissionAuthoringFields {
  mission_id: string;
  title: string;
  objective: string;
  success_criteria: string[];
  project_id?: string;
  context?: Record<string, unknown>;
  deliverables?: string[];
  research_phases?: Record<string, unknown>;
  tags?: string[];
  metadata?: Record<string, unknown>;
  research_depth?: ResearchDepth;
  status?: MissionStatus;
  created_by?: string;
}

export interface ApiMissionUpdate extends MissionAuthoringFields {
  title?: string;
  objective?: string;
  success_criteria?: string[];
  context?: Record<string, unknown>;
  deliverables?: string[];
  research_phases?: Record<string, unknown>;
  tags?: string[];
  metadata?: Record<string, unknown>;
  research_depth?: ResearchDepth;
  status?: MissionStatus;
  deepsearch_job_id?: string;
  result_document_ids?: string[];
  result_report_id?: string;
  result_markdown?: string;
  result_protocol?: Record<string, unknown>;
  error_message?: string;
  execution_metadata?: Record<string, unknown>;
}

export interface MissionListParams {
  page?: number;
  page_size?: number;
  status?: MissionStatus;
  project_id?: string;
}

export interface MissionSubmitResponse {
  status: string;
  mode: string;
  mission_id: string;
  uuid: string;
  message: string;
  job_id: string | null;
}

export interface ReportPromotionResponse {
  document_id: string;
  document_name: string;
  status: "processing" | "completed";
  message: string;
  chunk_count: number | null;
}

/** Compiled DeepSearch contract preview (T40.4). */
export interface MissionContractPreview {
  mission_id: string;
  mission_uuid: string;
  project_id?: string | null;
  named_entities: string[];
  objectives: Array<Record<string, unknown>>;
  evidence_slots: Array<Record<string, unknown>>;
  acceptance_checks: Array<Record<string, unknown>>;
  deliverable_schemas: Array<Record<string, unknown>>;
  coverage_thresholds: Record<string, number>;
  validation_thresholds: Record<string, number>;
}

File: app/schemas/mission.py
"""Pydantic schemas for mission entities.

Aligned with the Mission model from B16.1 with explicit DeepSearch-compatible fields.
Implements comprehensive validation per B16.3 spec.
"""

from __future__ import annotations

import re
from datetime import datetime
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator

# Valid mission statuses
MissionStatus = Literal[
    "draft", "queued", "in_progress", "completed",
    "blocked", "cancelled", "validation_failed",
]

# Valid research depth tiers
ResearchDepth = Literal["baseline", "deep", "alpha"]

# Mission ID pattern: starts with alphanumeric, can contain dots, dashes, underscores
MISSION_ID_PATTERN = re.compile(r"^[A-Za-z0-9][A-Za-z0-9._-]*$")


class MissionBase(BaseModel):
    """Shared attributes for mission operations."""

    mission_id: str = Field(
        ...,
        min_length=1,
        max_length=50,
        description="Human-readable mission identifier (e.g., B16.1)",
    )
    title: str = Field(
        ...,
        min_length=3,
        max_length=255,
        description="Mission title",
    )
    objective: str = Field(
        ...,
        min_length=10,
        description="What this mission aims to achieve (minimum 10 characters)",
    )
    success_criteria: list[str] = Field(
        ...,
        min_length=1,
        description="Array of measurable success conditions",
    )

    @field_validator("mission_id")
    @classmethod
    def validate_mission_id_format(cls, v: str) -> str:
        """Validate mission_id matches required pattern."""
        if not MISSION_ID_PATTERN.match(v):
            raise ValueError(
                "mission_id must start with alphanumeric and contain only "
                "letters, numbers, dots, dashes, or underscores"
            )
        return v

    @field_validator("success_criteria")
    @classmethod
    def validate_success_criteria(cls, v: list[str]) -> list[str]:
        """Validate success_criteria is non-empty and each item is a non-empty string."""
        if not v:
            raise ValueError("success_criteria must contain at least one item")
        for i, item in enumerate(v):
            if not isinstance(item, str):
                raise ValueError(f"success_criteria[{i}] must be a string")
            if not item.strip():
                raise ValueError(f"success_criteria[{i}] cannot be empty or whitespace")
        return v


class MissionCreate(MissionBase):
    """Payload for creating a mission."""

    project_id: UUID | None = Field(
        None,
        description="Project to associate this mission with",
    )
    context: dict[str, Any] | None = Field(
        default_factory=dict,
        description="Additional context object for the mission",
    )
    deliverables: list[str] | None = Field(
        default_factory=list,
        description="Array of expected deliverables",
    )
    research_phases: dict[str, Any] | None = Field(
        default_factory=dict,
        description="Research phase configuration",
    )
    tags: list[str] | None = Field(
        default_factory=list,
        description="Array of tags for categorization",
    )
    metadata: dict[str, Any] | None = Field(
        default_factory=dict,
        description="Arbitrary metadata object",
    )
    research_depth: ResearchDepth | None = Field(
        "baseline",
        description="Research depth tier. BASELINE (8-12 min, 50-60 sources): standard thorough research — use as default. DEEP (20-25 min, 30-40 vetted sources): stricter quality gates, min 5 loops. ALPHA (1+ hour, ~20 scrutinized sources): may reject if evidence insufficient.",
    )
    # Mission-authoring fields consumed by DeepSearch contract compiler (T40.1).
    background: str | None = Field(
        None,
        description="Free-form background prose orienting the research.",
    )
    focus: str | None = Field(
        None,
        description="Narrow framing for the research question.",
    )
    references: list[dict[str, Any]] | None = Field(
        None,
        description="Array of reference objects, each at minimum {title}.",
    )
    required_entities: list[str] | None = Field(
        None,
        description="Entities that MUST appear in the synthesized output.",
    )
    excluded_entities: list[str] | None = Field(
        None,
        description="Entities that MUST NOT appear in the synthesized output.",
    )
    expected_output_schema: dict[str, Any] | None = Field(
        None,
        description="DeepSearch OutputSchema describing the deliverable shape.",
    )
    coverage_thresholds: dict[str, Any] | None = Field(
        None,
        description="Dict of coverage gate thresholds applied during synthesis.",
    )
    validation_thresholds: dict[str, Any] | None = Field(
        None,
        description="Dict of validation gate thresholds applied during synthesis.",
    )
    deliverable_format: str | None = Field(
        None,
        description="Output rendering format hint (e.g. 'markdown report', 'comparison table').",
    )
    max_loops: int | None = Field(
        None,
        ge=1,
        description="Upper bound on DeepSearch research loop count.",
    )
    min_loops: int | None = Field(
        None,
        ge=1,
        description="Lower bound on DeepSearch research loop count.",
    )
    constraints: list[str] | None = Field(
        None,
        description="Constraint strings. Promoted from context['constraints'] in T40.1.",
    )
    status: MissionStatus | None = Field(
        "draft",
        description="Initial mission status",
    )
    created_by: str | None = Field(
        None,
        max_length=100,
        description="Agent or user who created this mission",
    )


class MissionUpdate(BaseModel):
    """Payload for updating a mission.

    All fields are optional - only provided fields will be updated.
    """

    title: str | None = Field(
        None,
        min_length=3,
        max_length=255,
        description="Mission title",
    )
    objective: str | None = Field(
        None,
        min_length=10,
        description="What this mission aims to achieve (minimum 10 characters)",
    )
    success_criteria: list[str] | None = Field(
        None,
        min_length=1,
        description="Array of measurable success conditions",
    )
    context: dict[str, Any] | None = Field(
        None,
        description="Additional context object for the mission",
    )
    deliverables: list[str] | None = Field(
        None,
        description="Array of expected deliverables",
    )
    research_phases: dict[str, Any] | None = Field(
        None,
        description="Research phase configuration",
    )
    tags: list[str] | None = Field(
        None,
        description="Array of tags for categorization",
    )
    metadata: dict[str, Any] | None = Field(
        None,
        description="Arbitrary metadata object",
    )
    research_depth: ResearchDepth | None = Field(
        None,
        description="Research depth tier. BASELINE (8-12 min, 50-60 sources): standard thorough research. DEEP (20-25 min, 30-40 vetted sources): stricter quality gates. ALPHA (1+ hour, ~20 scrutinized sources): may reject if evidence insufficient.",
    )
    # Mission-authoring fields (T40.1) — all optional on update.
    background: str | None = Field(
        None,
        description="Free-form background prose orienting the research.",
    )
    focus: str | None = Field(
        None,
        description="Narrow framing for the research question.",
    )
    references: list[dict[str, Any]] | None = Field(
        None,
        description="Array of reference objects, each at minimum {title}.",
    )
    required_entities: list[str] | None = Field(
        None,
        description="Entities that MUST appear in the synthesized output.",
    )
    excluded_entities: list[str] | None = Field(
        None,
        description="Entities that MUST NOT appear in the synthesized output.",
    )
    expected_output_schema: dict[str, Any] | None = Field(
        None,
        description="DeepSearch OutputSchema describing the deliverable shape.",
    )
    coverage_thresholds: dict[str, Any] | None = Field(
        None,
        description="Dict of coverage gate thresholds applied during synthesis.",
    )
    validation_thresholds: dict[str, Any] | None = Field(
        None,
        description="Dict of validation gate thresholds applied during synthesis.",
    )
    deliverable_format: str | None = Field(
        None,
        description="Output rendering format hint (e.g. 'markdown report').",
    )
    max_loops: int | None = Field(
        None,
        ge=1,
        description="Upper bound on DeepSearch research loop count.",
    )
    min_loops: int | None = Field(
        None,
        ge=1,
        description="Lower bound on DeepSearch research loop count.",
    )
    constraints: list[str] | None = Field(
        None,
        description="Constraint strings. Promoted from context['constraints'] in T40.1.",
    )
    status: MissionStatus | None = Field(
        None,
        description="Mission status",
    )
    deepsearch_job_id: str | None = Field(
        None,
        max_length=100,
        description="DeepSearch job ID for tracking async execution",
    )
    result_document_ids: list[UUID] | None = Field(
        None,
        description="Array of document UUIDs produced by this mission",
    )
    result_report_id: UUID | None = Field(
        None,
        description="Primary report generated from mission results",
    )
    result_markdown: str | None = Field(
        None,
        description="Raw markdown output from mission execution",
    )
    result_protocol: dict[str, Any] | None = Field(
        None,
        description="Mission Protocol compliant result object",
    )
    error_message: str | None = Field(
        None,
        description="Error details if mission failed",
    )
    execution_metadata: dict[str, Any] | None = Field(
        None,
        description="Execution metrics and debugging info",
    )

    @field_validator("success_criteria")
    @classmethod
    def validate_success_criteria(cls, v: list[str] | None) -> list[str] | None:
        """Validate success_criteria when provided."""
        if v is None:
            return v
        if len(v) == 0:
            raise ValueError(
                "success_criteria must contain at least one item if provided"
            )
        for i, item in enumerate(v):
            if not isinstance(item, str):
                raise ValueError(f"success_criteria[{i}] must be a string")
            if not item.strip():
                raise ValueError(f"success_criteria[{i}] cannot be empty or whitespace")
        return v


class MissionResponse(MissionBase):
    """Full mission representation returned from API."""

    id: UUID
    project_id: UUID | None = None
    project_name: str | None = Field(None, description="Name of the associated project")
    context: dict[str, Any] = Field(default_factory=dict)
    deliverables: list[str] = Field(default_factory=list)
    research_phases: dict[str, Any] = Field(default_factory=dict)
    tags: list[str] = Field(default_factory=list)
    metadata: dict[str, Any] = Field(default_factory=dict)
    research_depth: ResearchDepth | None = Field(
        "baseline",
        description="Research depth tier. BASELINE (8-12 min, 50-60 sources): standard thorough research. DEEP (20-25 min, 30-40 vetted sources): stricter quality gates. ALPHA (1+ hour, ~20 scrutinized sources): may reject if evidence insufficient.",
    )
    # Mission-authoring fields (T40.1).
    background: str | None = None
    focus: str | None = None
    references: list[dict[str, Any]] | None = None
    required_entities: list[str] | None = None
    excluded_entities: list[str] | None = None
    expected_output_schema: dict[str, Any] | None = None
    coverage_thresholds: dict[str, Any] | None = None
    validation_thresholds: dict[str, Any] | None = None
    deliverable_format: str | None = None
    max_loops: int | None = None
    min_loops: int | None = None
    constraints: list[str] | None = None
    status: MissionStatus
    queued_at: datetime | None = None
    started_at: datetime | None = None
    completed_at: datetime | None = None
    deepsearch_job_id: str | None = None
    execution_metadata: dict[str, Any] = Field(default_factory=dict)
    result_document_ids: list[UUID] = Field(default_factory=list)
    result_report_id: UUID | None = None
    result_markdown: str | None = None
    result_protocol: dict[str, Any] | None = None
    error_message: str | None = None
    created_at: datetime
    updated_at: datetime
    created_by: str | None = None

    model_config = ConfigDict(from_attributes=True)

    @field_validator("metadata", mode="before")
    @classmethod
    def map_mission_metadata(cls, v: Any, info) -> dict[str, Any]:
        """Map mission_metadata column to metadata field."""
        # Handle case where we get mission_metadata from ORM
        if v is None:
            # Try to get from the source object
            data = info.data if hasattr(info, "data") else {}
            return data.get("mission_metadata", {}) or {}
        return v or {}

    @field_validator("result_document_ids", mode="before")
    @classmethod
    def parse_document_ids(cls, v: Any) -> list[UUID]:
        """Parse document IDs from JSON storage."""
        if v is None:
            return []
        if isinstance(v, list):
            return [UUID(str(x)) if not isinstance(x, UUID) else x for x in v]
        return []

    @field_validator("constraints", mode="before")
    @classmethod
    def fallback_constraints_from_context(
        cls, v: Any, info
    ) -> list[str] | None:
        """Fallback to context['constraints'] when the column is null.

        T40.1 promoted `constraints` out of `context` into its own column.
        Missions authored before the migration only have it inside `context`;
        this keeps DeepSearch's existing reader working through the transition.
        An empty list is treated as an explicit author choice and preserved.
        """
        if v is not None:
            return v
        data = info.data if hasattr(info, "data") else {}
        ctx = data.get("context") if isinstance(data, dict) else None
        if isinstance(ctx, dict):
            legacy = ctx.get("constraints")
            if legacy:
                return legacy
        return v


# Alias for backwards compatibility
MissionRead = MissionResponse


class MissionSubmitResponse(BaseModel):
    """Response from submitting a mission for execution."""

    status: str = Field(..., description="Current mission status (queued)")
    mode: str = Field(..., description="Execution mode (worker or http)")
    mission_id: str = Field(..., description="Human-readable mission ID")
    uuid: UUID = Field(..., description="Mission UUID")
    message: str = Field(..., description="Status message")
    job_id: str | None = Field(None, description="DeepSearch job ID (http mode only)")


class MissionActionableError(BaseModel):
    """Structured error details for agent-facing mission workflows."""

    message: str = Field(..., description="Human-readable error message")
    mission_id: str | None = Field(
        None,
        description="Human-readable mission identifier when available",
    )
    uuid: UUID | None = Field(
        None,
        description="Mission UUID when available",
    )
    suggestion: str | None = Field(
        None,
        description="Concrete follow-up action to resolve the issue",
    )
    current_status: MissionStatus | None = Field(
        None,
        description="Current mission status, when relevant",
    )


class MissionErrorResponse(BaseModel):
    """HTTP error envelope used by mission endpoints."""

    detail: MissionActionableError


class ReportPromotionResponse(BaseModel):
    """Response from promoting a mission report to a document."""

    document_id: UUID = Field(..., description="UUID of the created document")
    document_name: str = Field(..., description="Name of the created document")
    status: str = Field(..., description="Processing status (processing or completed)")
    message: str = Field(..., description="Status message")
    chunk_count: int | None = Field(None, description="Number of chunks created")
